src/Features/LanguageServer/Protocol/RequestDispatcher.cs

```diff
@@ -0,0 +1,91 @@
+        {
+            var requestHandlerDictionary = ImmutableDictionary.CreateBuilder<string, (IRequestHandler, ILspMethodMetadata)>(StringComparer.OrdinalIgnoreCase);
+
+            // Create the actual request handlers from the providers.
+            var handlers = requestHandlerProviders.SelectMany(lazyProvider => lazyProvider.Value.CreateRequestHandlers());
+
+            // Store the request handlers in a dicitionary from method name (+ command name if workspace/executeCommand method) to handler instance.
+            foreach (var handler in handlers)
+            {
+                var lspMethodNameWithCommandName = handler.Metadata.MethodName + handler.Metadata.CommandName;
+                requestHandlerDictionary.Add(lspMethodNameWithCommandName, handler);
+            }
+
+            return requestHandlerDictionary.ToImmutable();
+        }
+
+        public Task<ResponseType> ExecuteRequestAsync<RequestType, ResponseType>(RequestExecutionQueue queue, string methodName, RequestType request, LSP.ClientCapabilities clientCapabilities,
+            string? clientName, CancellationToken cancellationToken) where RequestType : class
+        {
+            Contract.ThrowIfNull(request);
+            Contract.ThrowIfTrue(string.IsNullOrEmpty(methodName), "Invalid method name");
+
+            if (request is ExecuteCommandParams executeCommandRequest)
+            {
+                // If we have a workspace/executeCommand request, concat the command name with the method name
+                // to look up the appropriate request handler for the workspace/executeCommand's command name.
+                methodName += executeCommandRequest.Command;
+            }
+
+            var handlerEntry = _requestHandlers[methodName];
+            Contract.ThrowIfNull(handlerEntry, string.Format("Request handler entry not found for method {0}", methodName));
+
+            var mutatesSolutionState = handlerEntry.Metadata.MutatesSolutionState;
+
+            var handler = (IRequestHandler<RequestType, ResponseType>?)handlerEntry.RequestHandler;
+            Contract.ThrowIfNull(handler, string.Format("Request handler not found for method {0}", methodName));
+
+            return ExecuteRequestAsync(queue, request, clientCapabilities, clientName, methodName, mutatesSolutionState, handler, cancellationToken);
+        }
+
+        protected virtual Task<ResponseType> ExecuteRequestAsync<RequestType, ResponseType>(RequestExecutionQueue queue, RequestType request, ClientCapabilities clientCapabilities, string? clientName, string methodName, bool mutatesSolutionState, IRequestHandler<RequestType, ResponseType> handler, CancellationToken cancellationToken) where RequestType : class
+        {
+            return queue.ExecuteAsync(mutatesSolutionState, handler, request, clientCapabilities, clientName, methodName, cancellationToken);
+        }
+
+        internal TestAccessor GetTestAccessor()
```

Comment:
this is required to access the handlers' cache instances now that they are no longer mef exported.